clang/runtime/dpct-rt/include/kernel.hpp.inc

```diff
@@ -86,5 +126,237 @@ static kernel_function_info get_kernel_function_info(const void *function) {
+  unsigned char const *const ublob = reinterpret_cast<unsigned char const *const>(blob);
+  // Analyze DOS stub
+  if (ublob[0] != 0x4d ||
+      ublob[1] != 0x5a) {
+    throw std::runtime_error("Blob is not a Windows DLL.");
+  }
+  uint32_t pe_header_offset = extract32(ublob+0x3c);
+
+  ///////////////////////////////////////////////////////////////////////
+  // Ananlyze PE-header
+  unsigned char const *const pe_header = ublob + pe_header_offset;
+
+  // signature
+  uint32_t pe_signature = extract32(pe_header+0);
+  if (pe_signature!=0x00004550) {
+    throw std::runtime_error("PE-header signature is not 0x00004550");
+  }
+
+  // machine
+  uint16_t machine  = extract16(pe_header+4);
+  if (machine!=0x8664) {
+    throw std::runtime_error("Only DLLs for x64 supported");
+  }
+
+  // number of sections
+  uint16_t number_of_sections  = extract16(pe_header+6);
+
+  // sizeof optional header
+  uint16_t sizeof_optional_header = extract16(pe_header+20);
+
+  //magic
+  uint16_t magic = extract16(pe_header+24);
+  if (magic!=0x10b &&
+      magic!=0x20b) {
+    throw std::runtime_error("MAGIC is not 0x010b or 0x020b");
+  }
+
+  ///////////////////////////////////////////////////////////////////////
+  // Analyze tail of optional header
+  constexpr int coff_header_size = 24;
+
+  unsigned char const *const tail_of_optional_header = pe_header + coff_header_size + sizeof_optional_header;
+  if (extract64(tail_of_optional_header-8)!=0) {
+    throw std::runtime_error("Optional header not zero-padded");
+  }
+
+  ///////////////////////////////////////////////////////////////////////
+  // Analyze last section header
+  constexpr int section_header_size = 40;
+  unsigned char const *const last_section_header = tail_of_optional_header + section_header_size*(number_of_sections-1);
+
+  uint32_t sizeof_raw_data     = extract32(last_section_header+16);
+  uint32_t pointer_to_raw_data = extract32(last_section_header+20);
+
+  return sizeof_raw_data + pointer_to_raw_data;
+#else
+  if (blob[0] != 0x7F || blob[1] != 'E' || blob[2] != 'L' || blob[3] != 'F')
+    throw std::runtime_error("Blob is not in ELF format");
+
+  if (blob[4] != 0x02)
+    throw std::runtime_error("Only 64-bit headers are supported");
+
+  if (blob[5] != 0x01)
+    throw std::runtime_error("Only little-endian headers are supported");
+
+  unsigned char const *const ublob = reinterpret_cast<unsigned char const *const>(blob);
+  uint64_t e_shoff     = extract64(ublob+0x28);
+  uint16_t e_shentsize = extract16(ublob+0x3A);
+  uint16_t e_shnum     = extract16(ublob+0x3C);
+
+  return e_shoff + (e_shentsize * e_shnum);
+#endif
+}
+
+} // namespace detail
+
+class module {
+public:
+  module()          : ptr{nullptr} {}
+  module(void *ptr) : ptr{ptr}     {}
+
+  operator void *() const { return ptr; }
+
+void *get_ptr() {
+  return(ptr);
+}
+private:
+void *ptr;
+};
+
+namespace detail {
+
+static std::string mkfilename()
+{
```

Comment:
1. Why do we need to covert rawtime to string before hash? Can we hash directly with rawtime?
2. Please using c++ style string operation instead of sprintf() if needed.